File: 0515_Find_Largest_Value_in_Each_Tree_Row/solution.py
'''
515. Find Largest Value in Each Tree Row


You need to find the largest value in each row of a binary tree.

Example:
Input:

          1
         / \
        3   2
       / \   \
      5   3   9

Output: [1, 3, 9]

'''


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


INT_MIN = -(2 << 31)


class Solution(object):
    def largestValues(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        if not root:
            return res
        level = [root]
        while level:
            max_val = INT_MIN
            next_level = []
            for node in level:
                max_val = max(max_val, node.val)
                if node.left:
                    next_level.append(node.left)
                if node.right:
                    next_level.append(node.right)
            level = next_level
            res.append(max_val)
        return res

File: 0515_Find_Largest_Value_in_Each_Tree_Row/solution.cpp
/*
515. Find Largest Value in Each Tree Row

https://leetcode.com/problems/find-largest-value-in-each-tree-row/
*/

#include <vector>
using namespace std;

#include "common/TreeNode.h"
/* 
    Using row index
 */
class Solution {
public:
  vector<int> largestValues(TreeNode* root) {
    vector<int> ans;
    inorder(root, 0, ans);
    return ans;
  }
private:
  void inorder(TreeNode* root, int row, vector<int>& ans) {
    if (root == nullptr) return;
    while (ans.size() <= row) 
        ans.push_back(INT_MIN);    
    inorder(root->left, row + 1, ans);
    ans[row] = max(ans[row], root->val);
    inorder(root->right, row + 1, ans);
  }
};